crates/circuit/src/parameter_expression.rs

```diff
@@ -10,85 +10,1250 @@
+                    symbol_expr::Value::Complex(c) => ParameterValueType::Complex(c),
+                };
+            }
+        }
+        ParameterValueType::Expression(expr.clone())
+    }
+}
+
+#[pyfunction]
 fn _extract_value(value: &Bound<PyAny>) -> Option<ParameterExpression> {
-    if let Ok(i) = value.extract::<i64>() {
+    if let Ok(e) = value.extract::<ParameterExpression>() {
+        Some(e)
+    } else if let Ok(_) = value.extract::<String>() {
+        None
+    } else if let Ok(i) = value.extract::<i64>() {
         Some(ParameterExpression {
             expr: SymbolExpr::Value(symbol_expr::Value::from(i)),
+            uuid: 0,
+            qpy_replay: None,
+            parameter_symbols: None,
+            parameter_vector: None,
         })
     } else if let Ok(c) = value.extract::<Complex64>() {
+        if c.is_infinite() || c.is_nan() {
+            return None;
+        }
         Some(ParameterExpression {
             expr: SymbolExpr::Value(symbol_expr::Value::from(c)),
+            uuid: 0,
+            qpy_replay: None,
+            parameter_symbols: None,
+            parameter_vector: None,
         })
     } else if let Ok(r) = value.extract::<f64>() {
+        if r.is_infinite() || r.is_nan() {
+            return None;
+        }
         Some(ParameterExpression {
             expr: SymbolExpr::Value(symbol_expr::Value::from(r)),
+            uuid: 0,
+            qpy_replay: None,
+            parameter_symbols: None,
+            parameter_vector: None,
         })
-    } else if let Ok(s) = value.extract::<String>() {
-        if let Ok(expr) = parse_expression(&s) {
-            Some(ParameterExpression { expr })
-        } else {
-            None
-        }
     } else {
-        value.extract::<ParameterExpression>().ok()
+        None
+    }
+}
+
+// OP codes for QPY replay
+#[pyclass(sequence, module = "qiskit._accelerate.circuit")]
+#[derive(Clone, Debug)]
+pub enum _OPCode {
+    ADD = 0,
+    SUB = 1,
+    MUL = 2,
+    DIV = 3,
+    POW = 4,
+    SIN = 5,
+    COS = 6,
+    TAN = 7,
+    ASIN = 8,
+    ACOS = 9,
+    EXP = 10,
+    LOG = 11,
+    SIGN = 12,
+    GRAD = 13,
+    CONJ = 14,
+    SUBSTITUTE = 15,
+    ABS = 16,
+    ATAN = 17,
+    RSUB = 18,
+    RDIV = 19,
+    RPOW = 20,
+}
+
+// enum for QPY replay
+#[pyclass(sequence, module = "qiskit._accelerate.circuit")]
+#[derive(Clone, Debug)]
+pub enum OPReplay {
+    _INSTRUCTION {
+        op: _OPCode,
+        lhs: Option<ParameterValueType>,
+        rhs: Option<ParameterValueType>,
+    },
+    _SUBS {
+        binds: HashMap<ParameterExpression, ParameterValueType>,
+        op: _OPCode,
+    },
+}
+
+impl OPReplay {
+    pub fn new_instruction(
+        op: _OPCode,
+        lhs: Option<ParameterValueType>,
+        rhs: Option<ParameterValueType>,
+    ) -> OPReplay {
+        OPReplay::_INSTRUCTION { op, lhs, rhs }
+    }
+}
+
+// ===========================================
```

Comment:
Since these docstrings will end up in the documentation, these should probably copy what was previously the Python docs (see e.g. how it's done in the `SparseObservable` docs: https://github.com/Qiskit/qiskit/blob/a18e15163d8a93c0e072d4d3c50151a2cd9f036a/crates/accelerate/src/sparse_observable/mod.rs#L2412)
(same comment holds for the other Parameter objects that are exposed to Python)